Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs

```diff
@@ -1,46 +1,276 @@
-﻿using Content.Server.Atmos.Components;
+﻿using System.Linq;
+using Content.Server.Atmos;
+using Content.Server.Atmos.Components;
+using Content.Server.NodeContainer;
+using Content.Server.NodeContainer.Nodes;
 using Content.Server.Popups;
+using Content.Shared.Atmos;
+using Content.Shared.Atmos.Components;
 using Content.Shared.Interaction;
+using Content.Shared.Interaction.Events;
 using JetBrains.Annotations;
 using Robust.Server.GameObjects;
 using Robust.Shared.Player;
+using Robust.Shared.Utility;
+using static Content.Shared.Atmos.Components.SharedGasAnalyzerComponent;
 
 namespace Content.Server.Atmos.EntitySystems
 {
     [UsedImplicitly]
     public sealed class GasAnalyzerSystem : EntitySystem
     {
         [Dependency] private readonly PopupSystem _popup = default!;
+        [Dependency] private readonly AtmosphereSystem _atmo = default!;
+        [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
+        [Dependency] private readonly UserInterfaceSystem _userInterface = default!;
+
+        private HashSet<EntityUid> _activeAnalyzers = new();
+        private RemQueue<EntityUid> _invalidAnalyzers = new();
 
         public override void Initialize()
         {
             base.Initialize();
 
             SubscribeLocalEvent<GasAnalyzerComponent, AfterInteractEvent>(OnAfterInteract);
+            SubscribeLocalEvent<GasAnalyzerComponent, GasAnalyzerRefreshMessage>(OnRefreshMessage);
+            SubscribeLocalEvent<GasAnalyzerComponent, GasAnalyzerDisableMessage>(OnDisabledMessage);
+            SubscribeLocalEvent<GasAnalyzerComponent, ComponentShutdown>(OnComponentShutdown);
+            SubscribeLocalEvent<GasAnalyzerComponent, DroppedEvent>(OnDropped);
+            SubscribeLocalEvent<GasAnalyzerComponent, UseInHandEvent>(OnUseInHand);
         }
 
         public override void Update(float frameTime)
         {
-            foreach (var analyzer in EntityManager.EntityQuery<GasAnalyzerComponent>(true))
+            if (!_activeAnalyzers.Any())
+                return;
+
+            foreach (var analyzer in _activeAnalyzers)
             {
-                analyzer.Update(frameTime);
+                if (_invalidAnalyzers.List != null && _invalidAnalyzers.List.Contains(analyzer))
+                    continue;
+
+                if(!UpdateAnalyzer(analyzer))
+                    _invalidAnalyzers.Add(analyzer);
             }
+
+            foreach (var invalidAnalyzer in _invalidAnalyzers)
+            {
+                _activeAnalyzers.Remove(invalidAnalyzer);
+            }
+
+            _invalidAnalyzers.List?.Clear();
```

Comment:
with the active comp/entityquery stuff mentioned above, the `invalidAnalyzers` bookkeeping can be removed entirely by just calling `RemCompDeferred<ActiveGasAnalyzerComponent>` in the entityquery